keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,1044 @@
+ 
+ ...
+ 
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
+ 
+     // CDI devices for the container.
+     repeated CDIDevice CDI_devices = 17;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 18;
+ }
+ 
++// KubernetesResources contains the resource requests and limits as specified
++// in the Kubernetes core API ResourceRequirements.
++message KubernetesResources {
++    // Requests and limits from the Kubernetes container config.
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> requests = 1;
++    map<string, k8s.io.apimachinery.pkg.api.resource.Quantity> limits = 2;
++}
+```
+
+#### UpdateContainerResourcesRequest
+
+The UpdateContainerResourcesRequest message is extended to pass down unmodified
+resource requests from the PodSpec.
+
+```diff
+ message UpdateContainerResourcesRequest {
+     // ID of the container to update.
+     string container_id = 1;
+     // Resource configuration specific to Linux containers.
+     LinuxContainerResources linux = 2;
+     // Resource configuration specific to Windows containers.
+     WindowsContainerResources windows = 3;
+     // Unstructured key-value map holding arbitrary additional information for
+     // container resources updating. This can be used for specifying experimental
+     // resources to update or other options to use when updating the container.
+     map<string, string> annotations = 4;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 5;
+ }
+```
+
+#### PodSandboxConfig
+
+The PodSandboxConfig message (part of the RunPodSandbox request) will be
+extended to contain information about resources of all its containers known at
+the pod creation time. The container resources here are non-binding and only
+informational, e.g. for the runtime to prepare for optimal allocation of
+resources of all the containers of the Pod.
+
+```diff
+ message PodSandboxConfig {
+ 
+ ...
+ 
+     // Optional configurations specific to Linux hosts.
+     LinuxPodSandboxConfig linux = 8;
+     // Optional configurations specific to Windows hosts.
+     WindowsPodSandboxConfig windows = 9;
++
++    // Kubernetes resource spec of the containers in the pod.
++    PodResourceConfig pod_resources = 10;
+ }
+ 
++// PodResourceConfig contains information of all resources requirements of
++// the containers of a pod.
++message PodResourceConfig {
++    repeated ContainerResourceConfig init_containers = 1;
++    repeated ContainerResourceConfig containers = 2;
++}
+ 
++// ContainerResourceConfig contains information of all resource requirements of
++// one container.
++message ContainerResourceConfig {
++    // Name of the container
++    string name= 1;
++
++    // Kubernetes resource spec of the container
++    KubernetesResources kubernetes_resources = 2;
++
++    // Mounts for the container.
++    repeated Mount mounts = 3;
```

Comment:
Hi, @marquiz  I'm curious about the actual purpose of the newly added Mount here:
(1) what's the goal of adding the `Mount` ?
(2) Is there some new field Mount.Type, which I concern very much, added in it ?
In my case kata-containers/runtime-rs with direct-volume, I need a way to provide a Mount type and pass it to the OCI Spec Mount. I'm not sure if this field `Mount` can provide this ability?

details see：https://github.com/cncf-tags/container-device-interface/issues/162

Reply:
Hi @Apokleos, this new message (`ContainerResourceConfig`) is purely informational/"read-only". It's purpose is to pass down (to the CRI) information of all resources of all containers of the Pod at sandbox creation time, so that the runtime can prepare/reserve everything early. The `Mount` field also serves this purpose, setting up things correctly at vm/confidential container creation time. It is not meant for making any edits. The containers created later are expected to match with the information in this message.